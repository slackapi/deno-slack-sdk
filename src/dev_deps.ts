export { assertEquals } from "https://deno.land/std@0.134.0/testing/asserts.ts";
