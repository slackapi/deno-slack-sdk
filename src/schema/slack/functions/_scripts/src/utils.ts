import {
  ArrayFunctionProperty,
  FunctionProperty,
  FunctionRecord,
  FunctionsPayload,
  ObjectFunctionProperty,
} from "./types.ts";

const FUNCTIONS_JSON_PATH = "functions.json";

const green = "\x1b[92m";
const yellow = "\x1b[38;5;214m";
const red = "\x1b[91m";
const reset = "\x1b[0m";

export const greenText = (text: string) => green + text + reset;
export const yellowText = (text: string) => yellow + text + reset;
export const redText = (text: string) => red + text + reset;

export async function getSlackFunctions(
  functionsPayloadPath: string = FUNCTIONS_JSON_PATH,
): Promise<FunctionRecord[]> {
  const functionsPayload: FunctionsPayload = await Deno.readTextFile(
    functionsPayloadPath,
  ).then(JSON.parse);

  return functionsPayload.functions.filter((fn) => fn.type == "builtin");
}

export function isObjectFunctionProperty(
  property: FunctionProperty,
): property is ObjectFunctionProperty {
  return "properties" in property;
}

export function isArrayFunctionProperty(
  property: FunctionProperty,
): property is ArrayFunctionProperty {
  return "items" in property;
}
