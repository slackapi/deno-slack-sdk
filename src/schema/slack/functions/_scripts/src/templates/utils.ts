import { pascalCase } from "../../../../../../dev_deps.ts";
import { FunctionProperty, FunctionRecord } from "../types.ts";
import SchemaTypes from "../../../../../schema_types.ts";
import SlackTypes from "../../../../schema_types.ts";
import { InternalSlackTypes } from "../../../../types/custom/mod.ts";
import { AllowedTypeValue, AllowedTypeValueObject } from "./types.ts";
import { isCustomType } from "../../../../../../types/mod.ts";
import { isArrayFunctionProperty, isObjectFunctionProperty } from "../utils.ts";

// For coded workflow backwards compatibility, we alias certain builtins to their old names.
const ID_ALIASES: Record<string, string> = {
  "canvas_update_content_v2": "canvas_update_content",
};

function useIdAlias(callback_id: string): string {
  return callback_id in ID_ALIASES ? ID_ALIASES[callback_id] : callback_id;
}

export function autogeneratedComment(includeDate?: boolean): string {
  const dateString = includeDate ? ` on ${new Date().toDateString()}` : "";
  return `/** This file was autogenerated${dateString}. Follow the steps in src/schema/slack/functions/_scripts/README.md to rebuild **/`;
}

export function renderFunctionImport(callbackId: string): string {
  return `import ${getFunctionName(callbackId)} from "./${callbackId}.ts";`;
}

export function getFunctionName(callbackId: string): string {
  return pascalCase(useIdAlias(callbackId));
}

export function getSlackCallbackId(
  functionRecord: FunctionRecord,
): string {
  return `slack#/functions/${functionRecord.callback_id}`;
}

export function getParameterType(type: AllowedTypeValue): string {
  return isCustomType(type) ? type.id : type;
}

const getParameterList = (
  functionRecord: FunctionRecord,
): FunctionProperty[] => [
  ...functionRecord.input_parameters,
  ...functionRecord.output_parameters,
];

const hasTypeObject = (
  types: string[],
  typeObject: AllowedTypeValueObject,
): boolean =>
  types.some((t) =>
    Object.values(typeObject).map((val) => getParameterType(val)).includes(t)
  );

const extractTypes = (properties: FunctionProperty[]): string[] => {
  let types: Set<string> = new Set();
  properties.forEach((property) => {
    types.add(property.type);
    if (isArrayFunctionProperty(property)) {
      types = new Set([
        ...types,
        ...extractTypes([property.items]),
      ]);
    }
    if (isObjectFunctionProperty(property)) {
      types = new Set([
        ...types,
        ...extractTypes(Object.values(property.properties)),
      ]);
    }
  });
  return Array.from(types);
};

export function renderTypeImports(functionRecord: FunctionRecord) {
  const typescript: string[] = [];
  const functionRecordTypes = extractTypes(getParameterList(functionRecord));
  if (hasTypeObject(functionRecordTypes, SchemaTypes)) {
    typescript.push('import SchemaTypes from "../../schema_types.ts";');
  }
  if (hasTypeObject(functionRecordTypes, SlackTypes)) {
    typescript.push('import SlackTypes from "../schema_types.ts";');
  }
  if (hasTypeObject(functionRecordTypes, InternalSlackTypes)) {
    typescript.push(
      'import { InternalSlackTypes } from "../types/custom/mod.ts";',
    );
  }
  return typescript.join("\n");
}

export function sanitize(value: string): string {
  return JSON.stringify(value);
}
