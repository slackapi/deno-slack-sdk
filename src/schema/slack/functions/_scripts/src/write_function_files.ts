import {
  SlackFunctionModTemplate,
  SlackFunctionTemplate,
  SlackTestFunctionTemplate,
} from "./templates/mod.ts";
import { getSlackFunctions, greenText, redText } from "./utils.ts";
import { FunctionRecord } from "./types.ts";

const VALID_FILENAME_REGEX = /^[0-9a-zA-Z_\-]+$/;

async function main() {
  const slackFunctions: FunctionRecord[] = await _internals.getSlackFunctions();

  // Sorting alphabetically cause only a monster would generate these in a random order
  slackFunctions.sort((a, b) => a.callback_id.localeCompare(b.callback_id));

  await Promise.all(
    slackFunctions.map(async (functionRecord: FunctionRecord) => {
      console.log(
        `Generating code & tests for Slack function: ${
          greenText(functionRecord.callback_id)
        }`,
      );
      if (!VALID_FILENAME_REGEX.test(functionRecord.callback_id)) {
        console.log(
          `${redText("FAILURE:")} Invalid characters in callback_id: ${
            redText(functionRecord.callback_id)
          }`,
        );
        return;
      }
      const filename = `../${functionRecord.callback_id}.ts`;
      const testFilename = `../${functionRecord.callback_id}_test.ts`;

      const templateString = _internals.SlackFunctionTemplate(functionRecord);
      const templateTestString = _internals.SlackTestFunctionTemplate(
        functionRecord,
      );

      await _internals.writeTextFile(filename, templateString);
      await _internals.writeTextFile(testFilename, templateTestString);
    }),
  );

  console.log(
    `Generated ${slackFunctions.length} Slack functions with their unit tests`,
  );

  const modString = _internals.SlackFunctionModTemplate(slackFunctions);

  await _internals.writeTextFile("../mod.ts", modString);
  console.log("Updated functions module export");
}

export const _internals = {
  main,
  getSlackFunctions,
  SlackFunctionModTemplate,
  SlackFunctionTemplate,
  SlackTestFunctionTemplate,
  writeTextFile: (
    path: string,
    data: string,
    options?: Deno.WriteFileOptions,
  ): ReturnType<typeof Deno.writeTextFile> => {
    return Deno.writeTextFile(path, data, options);
  },
};

if (import.meta.main) {
  _internals.main();
}
