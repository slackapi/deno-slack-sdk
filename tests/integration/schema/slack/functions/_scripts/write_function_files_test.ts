import { assertSnapshot, mock } from "../../../../../../src/dev_deps.ts";
import { _internals } from "../../../../../../src/schema/slack/functions/_scripts/src/write_function_files.ts";
import { getSlackFunctions } from "../../../../../../src/schema/slack/functions/_scripts/src/utils.ts";

Deno.test(`write_function_files.ts ${_internals.main.name} function snapshot test`, async (t) => {
  const getSlackFunctionsStub = mock.stub(
    _internals,
    "getSlackFunctions",
    (
      functionsPayloadPath =
        "src/schema/slack/functions/_scripts/src/test/data/function.json",
    ) => {
      return getSlackFunctions(
        functionsPayloadPath,
      );
    },
  );

  const paths: string[] = [];
  const fileContents: string[] = [];

  const writeTextFileStub = mock.stub(
    _internals,
    "writeTextFile",
    // deno-lint-ignore require-await
    async (path, data) => {
      paths.push(path);
      fileContents.push(data);
    },
  );

  try {
    await _internals.main();
    mock.assertSpyCalls(writeTextFileStub, 3);
    mock.assertSpyCalls(getSlackFunctionsStub, 1);
    assertSnapshot(t, paths);
    assertSnapshot(t, fileContents);
  } finally {
    writeTextFileStub.restore();
    getSlackFunctionsStub.restore();
  }
});
