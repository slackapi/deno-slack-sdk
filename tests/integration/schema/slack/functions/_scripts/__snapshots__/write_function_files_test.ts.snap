export const snapshot = {};

snapshot[`write_function_files.ts main function snapshot test 1`] = `
[
  "../send_message.ts",
  "../send_message_test.ts",
  "../mod.ts",
]
`;

snapshot[`write_function_files.ts main function snapshot test 2`] = `
[
  "/** This file was autogenerated. Follow the steps in src/schema/slack/functions/README.md to rebuild **/
" +
    'import { DefineFunction } from "../../../functions/mod.ts";
' +
    'import SchemaTypes from "../../schema_types.ts";
' +
    'import SlackTypes from "../schema_types.ts";
' +
    "
" +
    'export default DefineFunction({callback_id: "slack#/functions/send_message",
' +
    'source_file: "",
' +
    'title: "Send a message to channel",
' +
    'description: "Send a message to channel",
' +
    "input_parameters: {properties: {channel_id: {type: SlackTypes.channel_id,
" +
    'description: "Search all channels",
' +
    'title: "Select a channel"},
' +
    "message: {type: SlackTypes.rich_text,
" +
    'description: "Add a message",
' +
    'title: "Add a message"},
' +
    "thread_ts: {type: SlackTypes.message_ts,
" +
    \`description: "Provide another message's ts value to make this message a reply",
\` +
    \`title: "Another message's timestamp value"},
\` +
    "object: {type: SchemaTypes.object,
" +
    'description: "Object type",
' +
    'title: "Object",
' +
    "properties: {event_type: {type: SchemaTypes.string},
" +
    "event_payload: {type: SchemaTypes.object}},
" +
    "additionalProperties: true,
" +
    'required: ["event_type","event_payload"]},
' +
    "test_array: {type: SchemaTypes.array,
" +
    'description: "test an array",
' +
    'title: "test array",
' +
    "items: {type: SchemaTypes.string}},
" +
    "interactive_blocks: {type: SlackTypes.blocks,
" +
    'description: "Button(s) to send with the message",
' +
    'title: "Button(s) to send with the message"}},
' +
    'required: ["channel_id","message"]},
' +
    "output_parameters: {properties: {message_ts: {type: SlackTypes.message_ts,
" +
    'description: "Message time stamp",
' +
    'title: "Message time stamp"},
' +
    "interactivity: {type: SlackTypes.interactivity,
" +
    'description: "Interactivity context",
' +
    'title: "interactivity"},
' +
    "message_context: {type: SlackTypes.message_context,
" +
    'description: "Reference to the message sent",
' +
    'title: "Reference to the message sent"}},
' +
    'required: ["message_ts","message_context"]}});',
  "/** This file was autogenerated. Follow the steps in src/schema/slack/functions/README.md to rebuild **/
" +
    'import { assertEquals, assertExists } from "../../../dev_deps.ts";
' +
    'import { DefineWorkflow } from "../../../workflows/mod.ts";
' +
    'import { ManifestFunctionSchema } from "../../../manifest/manifest_schema.ts";
' +
    'import SchemaTypes from "../../schema_types.ts";
' +
    'import SlackTypes from "../schema_types.ts";
' +
    'import SendMessage from "./send_message.ts";
' +
    "
" +
    'Deno.test("SendMessage generates valid FunctionManifest", () => {assertEquals(SendMessage.definition.callback_id, "slack#/functions/send_message");
' +
    'const expected: ManifestFunctionSchema = {source_file: "",
' +
    'title: "Send a message to channel",
' +
    'description: "Send a message to channel",
' +
    "input_parameters: {properties: {channel_id: {type: SlackTypes.channel_id,
" +
    'description: "Search all channels",
' +
    'title: "Select a channel"},
' +
    "message: {type: SlackTypes.rich_text,
" +
    'description: "Add a message",
' +
    'title: "Add a message"},
' +
    "thread_ts: {type: SlackTypes.message_ts,
" +
    \`description: "Provide another message's ts value to make this message a reply",
\` +
    \`title: "Another message's timestamp value"},
\` +
    "object: {type: SchemaTypes.object,
" +
    'description: "Object type",
' +
    'title: "Object",
' +
    "properties: {event_type: {type: SchemaTypes.string},
" +
    "event_payload: {type: SchemaTypes.object}},
" +
    "additionalProperties: true,
" +
    'required: ["event_type","event_payload"]},
' +
    "test_array: {type: SchemaTypes.array,
" +
    'description: "test an array",
' +
    'title: "test array",
' +
    "items: {type: SchemaTypes.string}},
" +
    "interactive_blocks: {type: SlackTypes.blocks,
" +
    'description: "Button(s) to send with the message",
' +
    'title: "Button(s) to send with the message"}},
' +
    'required: ["channel_id","message"]},
' +
    "output_parameters: {properties: {message_ts: {type: SlackTypes.message_ts,
" +
    'description: "Message time stamp",
' +
    'title: "Message time stamp"},
' +
    "interactivity: {type: SlackTypes.interactivity,
" +
    'description: "Interactivity context",
' +
    'title: "interactivity"},
' +
    "message_context: {type: SlackTypes.message_context,
" +
    'description: "Reference to the message sent",
' +
    'title: "Reference to the message sent"}},
' +
    'required: ["message_ts","message_context"]}};
' +
    "const actual = SendMessage.export();
" +
    "
" +
    "assertEquals(actual, expected);});
" +
    "
" +
    'Deno.test("SendMessage can be used as a Slack function in a workflow step", () => {const testWorkflow = DefineWorkflow({callback_id: "test_SendMessage_slack_function", 
' +
    'title: "Test SendMessage", 
' +
    'description: "This is a generated test to test SendMessage"});
' +
    'testWorkflow.addStep(SendMessage, {channel_id: "test",
' +
    'message: "test"});
' +
    "const actual = testWorkflow.steps[0].export();
" +
    "
" +
    'assertEquals(actual.function_id, "slack#/functions/send_message");
' +
    'assertEquals(actual.inputs, {channel_id: "test",
' +
    'message: "test"});});
' +
    "
" +
    'Deno.test("All outputs of Slack function SendMessage should exist", () => {const testWorkflow = DefineWorkflow({callback_id: "test_SendMessage_slack_function", 
' +
    'title: "Test SendMessage", 
' +
    'description: "This is a generated test to test SendMessage"});
' +
    'const step = testWorkflow.addStep(SendMessage, {channel_id: "test",
' +
    'message: "test"});
' +
    "assertExists(step.outputs.message_ts);
" +
    "assertExists(step.outputs.interactivity);
" +
    "assertExists(step.outputs.message_context);});",
  "/** This file was autogenerated on Fri Sep 08 2023. Follow the steps in src/schema/slack/functions/README.md to rebuild **/
" +
    'import SendMessage from "./send_message.ts";
' +
    "
" +
    "const SlackFunctions = {SendMessage} as const;
" +
    "
" +
    "export default SlackFunctions;",
]
`;
